File: Day197_2025/Q197_GFG.java
import java.util.*;

class Q197_GFG {
    static List<Integer> sieve(int n) {
        boolean[] isPrime = new boolean[n];
        Arrays.fill(isPrime, true);
        isPrime[0] = false;
        isPrime[1] = false;

        for (int i = 2; i < n; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j < n; j += i) {
                    isPrime[j] = false;
                }
            }
        }

        List<Integer> primes = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (isPrime[i]) {
                primes.add(i);
            }
        }
        return primes;
    }

    public static int countNumbers(int n) {
        int count = 0;
        List<Integer> primes = sieve((int)Math.sqrt(n) + 1);

        for (int i = 0; i < primes.size(); i++) {
            long p8 = 1L;
            for (int j = 0; j < 8; j++) p8 *= primes.get(i);
            if (p8 <= n) {
                count++;
            } else {
                break;
            }
        }

        for (int i = 0; i < primes.size(); i++) {
            for (int j = i + 1; j < primes.size(); j++) {
                long val = 1L * primes.get(i) * primes.get(i) * primes.get(j) * primes.get(j);
                if (val <= n) {
                    count++;
                } else {
                    break;
                }
            }
        }

        return count;
    }
}

File: Day197_2025/Q197_GFG.py
class Solution:
    def sieve(self, n):
        is_prime = [True] * n
        is_prime[0] = is_prime[1] = False
        for i in range(2, n):
            if is_prime[i]:
                for j in range(i * i, n, i):
                    is_prime[j] = False
        return [i for i, val in enumerate(is_prime) if val]

    def countNumbers(self, n):
        from math import isqrt
        count = 0
        primes = self.sieve(isqrt(n) + 1)

        for p in primes:
            if p ** 8 <= n:
                count += 1
            else:
                break

        for i in range(len(primes)):
            for j in range(i + 1, len(primes)):
                val = primes[i] ** 2 * primes[j] ** 2
                if val <= n:
                    count += 1
                else:
                    break

        return count
